### ISSUE TITLE
Missing tags selenium/<image-name>:4 in Docker Hub

### ISSUE THREAD

**Issue by fkrivsky** [SeleniumHQ/docker-selenium#1365]
## 🐛 Bug Report

We used the tag selenium/hub:4 (and nodes as well) to keep up with the latest release of grid 4.
However, recently the :4 tags disappeared from Docker Hub.
We discovered it when we had a failure after redeploying pre-existing yamls.
Is this intentional please? 

It's breaking our workflow in a couple of aspects:
1. [Diun](https://crazymax.dev/diun/) no longer notifies about image update
2. We cannot simply redeploy existing yamls which was easy and smooth.

## To Reproduce

Try to deploy docker image using previously working tag :4

## Expected behavior

Having :4 tag in Docker Hub that would denote latest image of version 4.x

## Environment

OS: CentOS, Docker Swarm and K3s Kubernetes
Docker version: 20.10.8
Exact Docker command to start the containers (if using docker-compose, provide 
the docker-compose file as well):
- Compose example:
```
version: '3.7'
services:
 selenium-hub:
    image: selenium/hub:4
    ports:
      - "4442:4442"
      - "4443:4443"
      - "4444:4444"
    environment:
      TZ: "America/New_York"
    deploy:
      labels:
        - "diun.enable=true"
      placement:
        constraints:
         - node.labels.grid == true

 chrome:
    image: selenium/node-chrome:4
    volumes:
      - /dev/shm:/dev/shm
    environment:
      - TZ=America/New_York
      - SE_EVENT_BUS_HOST=selenium-hub
      - SE_EVENT_BUS_PUBLISH_PORT=4442
      - SE_EVENT_BUS_SUBSCRIBE_PORT=4443
      - SCREEN_HEIGHT=1080
      - SCREEN_WIDTH=1920
      - SCREEN_DEPTH=24
    deploy:
      labels:
        - "diun.enable=true"
      replicas: 40
      placement:
        constraints:
         - node.labels.grid == true
      resources:
        limits:
          cpus: '1.0'
          memory: 800M
        reservations:
          # cpus: '0.25'
          memory: 100M
    entrypoint: bash -c 'SE_OPTS="--host $$HOSTNAME" /opt/bin/entry_point.sh'
```
- k8s manifest:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-hub
  labels:
    app: selenium-hub
    component: "selenium-grid-4"
spec:
  replicas: 1
  selector:
    matchLabels:
      app: selenium-hub
  template:
    metadata:
      labels:
        app: selenium-hub
      annotations:
        diun.enable: "true"
        diun.watch_repo: "false"
    spec:
      containers:
        - name: selenium-hub
          image: selenium/hub:4
          imagePullPolicy: Always
          ports:
            - containerPort: 4444
              protocol: TCP
            - containerPort: 4443
              protocol: TCP
            - containerPort: 4442
              protocol: TCP
          env:
            - name: TZ
              value: "America/New_York"
          resources:
            limits:
              memory: "1000Mi"
              cpu: ".5"
          livenessProbe:
            httpGet:
              path: /wd/hub/status
              port: 4444
            initialDelaySeconds: 30
            timeoutSeconds: 5
          readinessProbe:
            httpGet:
              path: /wd/hub/status
              port: 4444
            initialDelaySeconds: 30
            timeoutSeconds: 5
---
# Hub Service - expose via NodePort
apiVersion: v1
kind: Service
metadata:
  name: selenium-hub
  labels:
    component: "selenium-grid-4"
spec:
  selector:
    app: selenium-hub
  ports:
    - port: 4442
      targetPort: 4442
      nodePort: 32442
      protocol: TCP
      name: event-bus-publish-port
    - port: 4443
      targetPort: 4443
      nodePort: 32443
      protocol: TCP
      name: event-bus-subscribe-port
    - port: 4444
      targetPort: 4444
      nodePort: 32444
      protocol: TCP
      name: grid-webui-port
  type: NodePort
  sessionAffinity: None
---
# Chrome nodes
apiVersion: apps/v1
kind: Deployment
metadata:
  name: selenium-node-chrome
  labels:
    app: selenium-node-chrome
    component: "selenium-grid-4"
spec:
  replicas: 48
  selector:
    matchLabels:
      app: selenium-node-chrome
  # nodeSelector:
  #   selenium.grid: true
  template:
    metadata:
      labels:
        app: selenium-node-chrome
      annotations:
        diun.enable: "true"
        diun.watch_repo: "false"
    spec:
      volumes:
        - name: dshm
          emptyDir:
            medium: Memory
      containers:
        - name: selenium-node-chrome
          image: selenium/node-chrome:4
          imagePullPolicy: Always
          ports:
            - containerPort: 5555
              protocol: TCP
          volumeMounts:
            - mountPath: /dev/shm
              name: dshm
          env:
            - name: SE_EVENT_BUS_HOST
              value: "selenium-hub"
            - name: SE_EVENT_BUS_PUBLISH_PORT
              value: "4442"
            - name: SE_EVENT_BUS_SUBSCRIBE_PORT
              value: "4443"
            - name: TZ
              value: "America/New_York"
            - name: "SCREEN_HEIGHT"
              value: "1200"
            - name: "SCREEN_WIDTH"
              value: "1920"
            - name: "SCREEN_DEPTH"
              value: "24"
          resources:
            limits:
              memory: "1000Mi"
              cpu: ".5"
```

**Comment by diemol** [comment]
Yes, this was intentional, as using tags like that reduces transparency when new images are released. The common pattern is having users updating their Docker images without realising a new browser or Grid version is being used, this leads to unexpected errors and scenarios like "I did not change anything, now all my tests are failing". 

We want to motivate all users to declare full tags when using Docker images, as this would make them aware of the current state of their pipeline. 

The main motivation for this was this [issue](https://github.com/SeleniumHQ/docker-selenium/issues/1346), where a Docker image was released and this broke lots of pipelines around the world, mainly because `latest` or `3.141.59` was being used. 

We believe this will have a positive effect in the long term. However, this change is not set on stone and we are always welcome feedback from the community.

I will leave this issue open for a while in case others want to chime in.

**Comment by fkrivsky** [comment]
I see, thank you for the explanation.

I personally do not foresee any benefits - Diun is basically industry standard tool for image update notifications and it's very easy to set up. With it users can be always aware of what version is available.
Now, there is no single tag for Diun to watch so it breaks this usage. Using _diun.watch_repo=true_ option makes too many docker hub calls and triggers adverse response, banning us from docker hub calls intermittently. 
Having to watch for updates manually is a step back in our efforts to automate infrastructure setup...

Said that, I do understand that you are on the receiving end with problems faced by less responsible users...

**Comment by nickgaya** [comment]
Hi, since #1369 was closed as a duplicate of this ticket I'd like to reiterate the main points here.

* This issue was caused by commit 169dc4a for issue #1360.
* That issue is based on the assertion that floating tags are a misuse of tagging and should be removed.
* According to the Docker [documentation](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier), tags are intended to be mutable, while image digests are immutable identifiers.
* The change to remove rolling tags breaks workflows for users that want to constrain the image version without fully specifying it.

**Comment by nickgaya** [comment]
Addressing some of the points @diemol brought up in their comment:

> The common pattern is having users updating their Docker images without realising a new browser or Grid version is being used, this leads to unexpected errors and scenarios like "I did not change anything, now all my tests are failing". We want to motivate all users to declare full tags when using Docker images, as this would make them aware of the current state of their pipeline.

I think this could be better addressed through improved documentation so that users are empowered to choose the appropriate tag based on their use-case.

Users who favor stability over timely updates can use fixed tags or image digests, while users who want to run their tests on the latest available version can consciously accept the risk that their pipelines may occasionally break due to an update or regression.

> The main motivation for this was [this issue](https://github.com/SeleniumHQ/docker-selenium/issues/1346), where a Docker image was released and this broke lots of pipelines around the world, mainly because latest or 3.141.59 was being used.

If I understand correctly, changing tagging wouldn't have avoided the issue — the problematic image would have been published either way.  Without floating tags, the problem would still emerge eventually as users updated their pipelines to the new version.

Adding some basic acceptance tests for the docker images before publishing could help reduce the likelihood of major regressions.

**Comment by diemol** [comment]
TL;DR;
I reverted the change and the tags were pushed again. https://github.com/SeleniumHQ/docker-selenium/releases/tag/4.0.0-rc-2-prerelease-20210907

Long version:

First, answering some comments from above:
>According to the Docker [documentation](https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier), tags are intended to be mutable, while image digests are immutable identifiers.

I believe that is an interpretation of that text, I do not see where the Docker docs actually mention that explicitly. I perceive this is more the decision of a project to run their tags. In our case, I felt it is more accurate to use immutable tags because more than one piece can move in a release. Using a mutable tag can mean that the browser version, and driver and more are different. Troubleshooting issues when a complete tag is not given is a pain. I wish more people could find time to help maintain the project.

>I think this could be better addressed through improved documentation so that users are empowered to choose the appropriate tag based on their use-case.

The docs all point to an "exact" tag. But yeah, there is always room for improvement. Contributions for the readme or the docs are always accepted.

>If I understand correctly, changing tagging wouldn't have avoided the issue — the problematic image would have been published either way. Without floating tags, the problem would still emerge eventually as users updated their pipelines to the new version.

It would have avoided the issue, because users would have needed to explicitly change the tag and then notice the issue, and then they could have rolled back the change. This simply makes users be more in control of their pipelines. Yes, the Docker image would still have the issue, but users would be aware of that instead of being surprised by the issue.

>Adding some basic acceptance tests for the docker images before publishing could help reduce the likelihood of major regressions.

There is already a set of tests executed before pushing the images to Docker Hub. Contributions to improve those tests and to check them is also welcome.


--- 

I think users should be aware of the tags they use, not knowing what browser and Selenium version is being used to run tests should be avoided, as this could simple make test debugging and troubleshooting even harder.
